### DISCUSSION TITLE
How use pycharm debugger mode to develop my fastapi server

### DISCUSSION THREAD

**Discussion by HalfLeaf** [fastapi/fastapi#8018] [Questions] [answered]
Like this, I can run my fastapi code! But I can't run it in debug mode for Pycharm IDE !
![image](https://user-images.githubusercontent.com/34050019/69472122-4682c780-0de1-11ea-852a-765013bbfe8e.png)

So,  How use pycharm debugger mode to develop my fastapi server?

**Comment by dmontagu** [comment]
The easiest way is to make calls to your ASGI app using the starlette `TestClient`. You can see examples in the fastapi tests.

**Comment by coryvirok** [comment, marked as answer]
The easiest way is to just add a `__main__` section to your app and run uvicorn by hand. Described here - https://fastapi.tiangolo.com/tutorial/debugging/

```py
if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

**Comment by BinNong** [comment]
> Like this, I can run my fastapi code! But I can't run it in debug mode for Pycharm IDE !
> ![image](https://user-images.githubusercontent.com/34050019/69472122-4682c780-0de1-11ea-852a-765013bbfe8e.png)
> 
> So, How use pycharm debugger mode to develop my fastapi server?

Hi @HalfLeaf，  I also encounter this problem. Have you solved it ?

**Comment by BinNong** [comment]
> > Like this, I can run my fastapi code! But I can't run it in debug mode for Pycharm IDE !
> > ![image](https://user-images.githubusercontent.com/34050019/69472122-4682c780-0de1-11ea-852a-765013bbfe8e.png)
> > So, How use pycharm debugger mode to develop my fastapi server?
> 
> Hi @HalfLeaf， I also encounter this problem. Have you solved it ?

I have sovled this problem. See: https://fastapi.tiangolo.com/tutorial/debugging/

**Comment by HalfLeaf** [comment]
> ```python
>  uvicorn.run（app，host = “ 0.0.0.0 ”，port = 8000）
> ```

yes, you are right，great， thank you！

**Comment by tiangolo** [comment]
Thanks for the help here everyone!

Thanks @HalfLeaf for reporting back and closing the issue :+1:

**Comment by dudil** [comment]
Great help!

@HalfLeaf
Just small addition that you might find helpful.
In order to support Uvicorn reload of code change (very useful when debugging!) update the code into:
``` python
if __name__ == "__main__":
    uvicorn.run("main:app", host="127.0.0.1", port=8000, reload=True)
```
Side note: *main* is assuming your python module name is main.py, as it mentioned on the (excellent!) documentation - https://fastapi.tiangolo.com/tutorial/first-steps/

(@tiangolo - maybe update also https://fastapi.tiangolo.com/tutorial/debugging/ ?)

**Comment by AndreMPCosta** [comment]
Well I've tried the sample code in [https://fastapi.tiangolo.com/tutorial/debugging/](url)

![image](https://user-images.githubusercontent.com/3736805/139041503-71f6e88a-7ac4-4d81-a43e-e877e5b45a6f.png)

![image](https://user-images.githubusercontent.com/3736805/139041615-e8074e6f-0e45-4e3f-8184-f41173b1fa8d.png)

![image](https://user-images.githubusercontent.com/3736805/139041913-4bf535ee-929d-4565-adc5-57f3c44594e2.png)

It does not trigger the breakpoint.

**Comment by lanzani** [comment]
> Well I've tried the sample code in [https://fastapi.tiangolo.com/tutorial/debugging/](url)
> 
> ![image](https://user-images.githubusercontent.com/3736805/139041503-71f6e88a-7ac4-4d81-a43e-e877e5b45a6f.png)
> 
> ![image](https://user-images.githubusercontent.com/3736805/139041615-e8074e6f-0e45-4e3f-8184-f41173b1fa8d.png)
> 
> ![image](https://user-images.githubusercontent.com/3736805/139041913-4bf535ee-929d-4565-adc5-57f3c44594e2.png)
> 
> It does not trigger the breakpoint.

Hei! I've just started learning FastAPI and I have the same issue... Did you figure out why this doesn't work?

**Comment by benoit-bnp** [comment]
I'm facing the same issue, any update on that ?

Edit: it works when `reload=False`

**Comment by bcygni** [comment]
For me, I was able to get the debugger working by unchecking the **Gevent compatible** option in the Python Debugger settings. I believe this is turned off by default but I had turned it on previously so I could debug some locust tests. 

![image](https://github.com/tiangolo/fastapi/assets/26200721/ac0012aa-f90a-4498-b64e-3de10f366c58)
